File: medium/994-Rotting-Oranges/BFS.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        
        Set<Integer> freshes = new HashSet<>();
        Queue<Integer> queue = new LinkedList<>();
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    freshes.add(i * n + j);
                }
                else if (grid[i][j] == 2) {
                    queue.offer(i * n + j);
                }
            }
        }
        
        int counter = 0;
        int[][] directions = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
        
        while (!queue.isEmpty() && !freshes.isEmpty()) {
            int numIter = queue.size();
            for (int k = 0; k < numIter; k++) {
                int val = queue.poll();
                int i = val / n;
                int j = val % n;
                for (int[] dir: directions) {
                    int[] newPos = {i + dir[0], j + dir[1]};
                    if (newPos[0] < 0 || newPos[0] >= m || newPos[1] < 0 || newPos[1] >= n) {
                        continue;
                    }                    
                    int newVal = newPos[0] * n + newPos[1];
                    if (freshes.contains(newVal)) {
                        freshes.remove(newVal);
                        queue.offer(newVal);
                    }
                }
            }
            ++counter;
        }
        
        return freshes.isEmpty() ? counter : -1;        
    }
}

File: medium/200-Number-of-Islands/BFS.kt
class Solution {
    fun numIslands(grid: Array<CharArray>): Int {
        val m = grid.size
        val n = grid[0].size
        
        var ans = 0
        var visited = Array(m) {BooleanArray(n) {false}}
        
        for (i in 0..(m - 1)) {
            for (j in 0..(n - 1)) {
                if ((grid[i][j] == '0') || visited[i][j]) continue;
                ++ans
                var queue = ArrayDeque<Int>()
                queue.offer(i * n + j)
                
                while (!queue.isEmpty()) {
                    val crr = queue.poll()
                    val x = crr / n
                    val y = crr % n
                    
                    if ((x > 0) && (grid[x - 1][y] == '1') && !visited[x - 1][y]) {
                        visited[x - 1][y] = true
                        queue.offer((x - 1) * n + y)
                    }
                    
                    if ((x < m - 1) && (grid[x + 1][y] == '1') && !visited[x + 1][y]) {
                        visited[x + 1][y] = true
                        queue.offer((x + 1) * n + y)
                    }
                    
                    if ((y > 0) && (grid[x][y - 1] == '1') && !visited[x][y - 1]) {
                        visited[x][y - 1] = true
                        queue.offer(x * n + y - 1)
                    }
                    
                    if ((y < n - 1) && (grid[x][y + 1] == '1') && !visited[x][y + 1]) {
                        visited[x][y + 1] = true
                        queue.offer(x * n + y + 1)
                    }

                }
            }
        }
        
        return ans
        
    }
}
